seed/seed20.go

```diff
@@ -0,0 +1,348 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package seed
+
+/* ATTN this should *not* use:
+
+* dirs package: it is passed an explicit directory to work on
+
+* release.OnClassic: it assumes classic based on the model classic
+  option; consistency between system and model can/must be enforced
+  elsewhere
+
+*/
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/asserts/snapasserts"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/seed/internal"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/timings"
+)
+
+type seed20 struct {
+	systemDir string
+
+	db asserts.RODatabase
+
+	model *asserts.Model
+
+	snapDeclsByID   map[string]*asserts.SnapDeclaration
+	snapDeclsByName map[string]*asserts.SnapDeclaration
+
+	snapRevsByID map[string]*asserts.SnapRevision
+
+	auxInfos map[string]*internal.AuxInfo20
+
+	snaps             []*Snap
+	essentialSnapsNum int
+}
+
+func (s *seed20) LoadAssertions(db asserts.RODatabase, commitTo func(*asserts.Batch) error) error {
+	if db == nil {
+		// a db was not provided, create an internal temporary one
+		var err error
+		db, commitTo, err = newMemAssertionsDB()
+		if err != nil {
+			return err
+		}
+	}
+
+	assertsDir := filepath.Join(s.systemDir, "assertions")
+	// collect assertions that are not the model
+	var declRefs []*asserts.Ref
+	var revRefs []*asserts.Ref
+	checkAssertion := func(ref *asserts.Ref) error {
+		switch ref.Type {
+		case asserts.ModelType:
+			return fmt.Errorf("system cannot have any model assertion but the one in the system model assertion file")
+		case asserts.SnapDeclarationType:
+			declRefs = append(declRefs, ref)
+		case asserts.SnapRevisionType:
+			revRefs = append(revRefs, ref)
+		}
```

Comment:
Silly question, what about all other assertion types? They won't be encountered in this context or are simply irrelevant here?

Reply:
there can be more assertions types, but we keep snap-declaration and snap-revision to make maps below, and we use the maps in lookupVerifiedRevision to go:

snap-id or snap-name -> snap-declaration (for the snap ame if we started from snap-id) -> snap-revision -> snap revision